core/multigrid/pgm.cpp

```diff
@@ -227,10 +237,238 @@ void Pgm<ValueType, IndexType>::generate()
+                        ->get_communicator();
+        auto num_rank = comm.size();
+        // Only support csr matrix currently.
+        auto pgm_local_op = std::dynamic_pointer_cast<const csr_type>(
+            matrix->get_local_matrix());
+        // If system matrix is not csr or need sorting, generate the csr.
+        if (!parameters_.skip_sorting || !pgm_local_op) {
+            pgm_local_op = convert_to_with_sorting<csr_type>(
+                exec, matrix->get_local_matrix(), parameters_.skip_sorting);
+        }
+        auto result = this->generate_local(pgm_local_op);
+        auto non_local_matrix = matrix->get_non_local_matrix();
+        auto non_local_size = non_local_matrix->get_size()[1];
+        array<IndexType> non_local_agg(exec, non_local_size);
+        // prolong, restrict, only needs the local infomation
+        // get agg information (prolong_row_gather row idx)
+        communicate(matrix, agg_, non_local_agg);
+        // generate non_local_col_map
+        non_local_agg.set_executor(exec->get_master());
+        array<IndexType> non_local_col_map(exec->get_master(), non_local_size);
+        array<IndexType> part_id(exec->get_master(), non_local_size);
+        array<IndexType> index(exec->get_master(), non_local_size);
+        auto recv_offsets = matrix->get_recv_offsets();
+        for (int i = 0; i + 1 < recv_offsets.size(); i++) {
+            for (auto j = recv_offsets.at(i); j < recv_offsets.at(i + 1); j++) {
+                part_id.get_data()[j] = i;
+                index.get_data()[j] = j;
+            }
+        }
+        // do it in host currently.
+        auto it = detail::make_zip_iterator(
+            part_id.get_data(), non_local_agg.get_data(), index.get_data());
+        // prepare tuple <part_id, local_agg, index>
+        // sort by <part_id, local_agg> or did segment sort
+        std::sort(it, it + non_local_size);
+        // add additional in tail such that the offset easily handle it.
+        array<IndexType> renumber(exec->get_master(), non_local_size + 1);
+        renumber.get_data()[0] = 0;
+        // renumber (prefix_sum) with not eqaul <part_id, local_agg>
+        for (int i = 1; i < non_local_size; i++) {
+            if (part_id.get_data()[i] != part_id.get_data()[i - 1] ||
+                non_local_agg.get_data()[i] !=
+                    non_local_agg.get_data()[i - 1]) {
+                renumber.get_data()[i] = renumber.get_data()[i - 1] + 1;
+            } else {
+                renumber.get_data()[i] = renumber.get_data()[i - 1];
+            }
+        }
+        renumber.get_data()[non_local_size] =
+            renumber.get_data()[non_local_size - 1] + 1;
+        // create col map
+        // for each thread i, col_map[tuple[i].index] = map[i]
+        for (int i = 0; i < non_local_size; i++) {
+            non_local_col_map.get_data()[index.get_data()[i]] =
+                renumber.get_data()[i];
+        }
+        // get new recv_size and recv_offsets
```

Comment:
I think you can also easily compute the send_sizes/offsets here, instead of communicating them.